### DISCUSSION TITLE
[Tutorial] How to delete the sector correctly

### DISCUSSION THREAD

**Discussion by de-authority** [filecoin-project/lotus#6950] [Tutorials]
# How to properly delete sectors
Before you delete a sector please make sure that you have taken the necessary rescue measures, such as encountering storage failure, network failure, scheduling failure, etc. you should go through a series of debugging, troubleshooting , and finally consider deleting the sector, carefully delete the sector, do not forget that the command has an option --really-do-it
 
## 1: How to delete the sectors in packing,PreCommitFailed and SealPreCommit1Failed status?

These sectors can be deleted directly with the following command because they are not yet pledged. All sectors before the completion of precommit phase can be deleted by this method.

```sh
 $ lotus-miner sectors remove --really-do-it <sectorId>
```
##### 1:First you should try the following command to delete
```sh
lotus-miner sealing abort <JobId>
```
```sh
lotus-miner sectors remove <SectorId>
```
If the above two commands fail to remove, you can perform the following steps
##### 2. Find a free worker (if all workers are working full, then you can remove it by executing
```sh
lotus-worker tasks disable [command options] [UNS|C2|PC2|PC1|AP]
```
Disable all tasks, wait for the worker to finish executing the current task, there will be no more  tasks, Then recover with the following command
```sh
lotus-worker tasks enable [command options] [UNS|C2|PC2|PC1|AP]
```
create empty sector files with corresponding sector IDs in the unsealed and sealed directories of the LOTUS_WORKER_PATH of the worker, assuming the sector ID is 100
Then, in the unsealed and sealed directories respectively, execute.
```sh
touch s-t0xxxxxx-100# where t0xxxxxx is the miner ID
```
##### 4. After the sector starts to sealing and you can see it in the lotus-miner sealing jobs list, you can execute the command to terminate the task first.

```sh 
lotus-miner sealing abort <JobId>
```
##### 5: Execute
```sh 
lotus-miner sectors remove <SectorId>
```
## 3: What if the delete status is Committing and keeps getting stuck on sectors with those statuses?  
> Note that deleting any sectors where precommit has completed will result in a collateral forfeit during the prcommit phase, [reference documentation](https://spec.filecoin.io/#section-systems.filecoin_mining.sector.adding_storage)  
> PreCommitting a Sector: A Miner publishes a Sector's SealedCID, through miner.PreCommitSector of miner.PreCommitSectorBatch, and makes a deposit. The Sector is now registered to the Miner, and the Miner must ProveCommit the Sector or lose their deposit.  
##### 2: After creating and restarting the Worker, the Sector will not appear in the list of lotus-miner sealing jobs, but will directly become CommitFailed, at this time, execute the following command to delete the Sector
```shell 
lotus-miner sectors remove <SectorId>
```

## 4: It is not recommended to remove commitfaied. finalizedfailed and all sectors that completed precommit or completed provcommit
It is recommended to restore the error sector to the state before the error and redo this step for the error
- Example: sector commitfailed. At this point pre-commit has been completed, reset the sector state to commit1
## 5: How to delete a sector that cannot be recovered because of a storage failure
Delete sectors must be deleted on the chain first and then locally, so as to minimize the loss, please remember the order of execution, this is quite important
 
##### 1: Execution

Delete sectors must be chained first and then deleted locally to minimize the loss.
 
```shell 
lotus-miner sectors terminate --really-do-it <sectorNum>
```
The main purpose of this step is to clear the data on the chain and minimize the penalty
##### 3: Execute

```shell 
lotus-miner sectors remove --really-do-it <sectorNum>
```
The main effect of this step is to clear the storage
 

## 6: Remove the most common errors in sectors

Many people will directly execute: lotus-miner sectors remove --really-do-it <sectorNum>, which is wrong. Then they can't find the sector number and can't terminate. In this case, you can remove the sectors by
```shell 
lotus-miner sectors list --fast --states Remomved
```
to see the sector number, and then run
```shell
lotus-miner sectors terminate --really-do-it <sectorNum> 
```
 
Thanks  @tears and @[bugxzhu](https://github.com/bugxzhu) for their help

**Comment by stuberman** [comment]
Are you certain this is good advice? I believe that before a sector puts a message on chain such as PreCommit, you can/should use **remove**, but once you put a message on chain, you should really use **terminate** to reduce your penalty.

**Reply by de-authority** [reply]
@stuberman yeah. i think i do the same thing as you  said in the tutorial

**Reply by de-authority** [reply]
@stuberman which part of my tutorial mislead you?

**Reply by stuberman** [reply]
The question is how soon can you use terminate, can it be used during the **SubmitPreCommitBatch** task?

**Reply by de-authority** [reply]
Ｉdon't think you can use terminate during SubmitPreCommitBatch.  that's another problem when you get stuck in SubmitPreCommitBatch

**Comment by Fatman13** [comment]
If you are using venus, replace `lotus-miner` with `venus-sealer`.
